Fix: Enhanced Product Collection's Price Range filter by improving currency parsing logic to discard unwanted characters

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR addresses an issue in the PriceTextField component, highlighted by @dinhtungdu in [this github comment](https://github.com/woocommerce/woocommerce/issues/45315#issuecomment-1985138515), where the input field would clear itself if users omitted a space between the numeric value and the currency symbol when the currency is configured to appear on the right with a space (e.g., "20$" instead of "20 $"). The problem was rooted in the strict parsing logic that did not account for variations in user input related to currency formatting.

The fix involves an enhancement to the numeric parsing function, which now effectively handles various user inputs regardless of space or placement of the currency symbol. The revised parsing logic uses a regular expression to strip out any characters that are not numeric or the designated decimal separator, and then replaces the locale-specific decimal separator with a period for standard numeric conversion. This approach not only fixes the specific issue but also improves the robustness of the component against similar input variations.

Closes #47351 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new post
2. Add Product Collection block & choose "Product Catalog" collection
3. Add "Price Range" filter in Inspector controls
4. Test Min & Max inputs of Price Filter. Verify that there is no issue. Here are some example scenarios to test

**Case 1:**
1. In Currency Options, set currency position to "Right with space" as shown in screenshot below:
![image](https://github.com/woocommerce/woocommerce/assets/16707866/95b1a24a-dedc-4cc1-ba42-3717642ca2df)
2. Add any value to "Min" field of Price Range filter, for example 20 and click away. You will notice that it will have a space between value & currency symbol i.e. "20 $". This is as expected. 
3. Now remove the space between value & currency symbol i.e. "20$" and click away. Verify that it should become "20 $".
	- Previously, the input field was getting cleared in this case, which was an issue. 
4. Publish the post and verify the only products within selected Price Range are shown on the Frontend. 

**Case 2:**
Try entering a MIN or MAX price with a decimal, for example, "10.123764". After clicking away, number of decimals should become equal to as set in "Currency Options" as you can see in screenshot above. In my case, It should become "10.12 $"

**Case 3:**
Try entering random values. Here are some inputs and expected outputs:
- 10ABCD -> 10 $
- 10000 -> 10,000 $ (Note, how it's using a Thousand separator from Currency Options setting)
- 1234 KJASHDH -> 1,234 $

As you can see above, it will try to parse numeric values from the input and discard unwanted characters like alphabets, space, etc. 

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Fix: Enhanced Product Collection's Price Range filter by improving currency parsing logic to discard unwanted characters

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

imanish003 committed:
Fix input clearing issue in PriceTextField with improved currency parsing

This commit addresses a usability issue in the PriceTextField component, highlighted by @dinhtungdu, where the input field would clear itself if users omitted a space between the numeric value and the currency symbol when the currency is configured to appear on the right with a space (e.g., "20$" instead of "20 $"). The problem was rooted in the strict parsing logic that did not account for variations in user input related to currency formatting.

The fix involves an enhancement to the numeric parsing function, which now effectively handles various user inputs regardless of space or placement of the currency symbol. The revised parsing logic uses a regular expression to strip out any characters that are not numeric or the designated decimal separator, and then replaces the locale-specific decimal separator with a period for standard numeric conversion. This approach not only fixes the specific issue but also improves the robustness of the component against similar input variations.

imanish003 committed:
Refactor currency parsing and improve handling in PriceTextField

1. Renaming `formatValueAsCurrency` to `formatNumberAsCurrency` to better
   reflect its functionality.
2. Moving the `convertCurrencyStringToNumber` function outside of the
   PriceTextField component, enhancing readability and reusability.
3. Changing the internal state handling from `newValue` to `inputValue`,
   directly managing the string input and parsing it as needed for further
   processing. This will force input field to re-render.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

kmanijak reviewed (approved):
Tested and parsing works as expected!

There's one typo in testing steps:
> Now remove the space between value & currency symbol i.e. "20$" and click away. Verify that it should become "10 $".

Should be:

> Now remove the space between value & currency symbol i.e. "20$" and click away. Verify that it should become "20 $".

---

Also, I found one minor issue. The default input value is empty (placeholder says `Auto`). But when you interact with input and remove the value, then it defaults to `0` which is kind of misleading:
![image](https://github.com/woocommerce/woocommerce/assets/20098064/c6695f9f-5a88-42dc-89f0-c1cfa690a64c)

In this image, there's MIN set to `0,01` and MAX set to nothing but displays `0`. But I double checked it can be also reproduced in `trunk` so just raising for clarity! Otherwise approving!

imanish003 commented:
> There's one typo in testing steps:

I fixed it, thanks for pointing it out 🙏🏻

> Also, I found one minor issue. The default input value is empty (placeholder says `Auto`). But when you interact with input and remove the value, then it defaults to `0` which is kind of misleading:

Good catch! I am looking into it. I can probably fix it with this PR as well.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into fix/47351-product-collection-improve-handling-of-currency-input-formatting-in-price-range-filter

imanish003 committed:
Fix: Handle empty string in price conversion to prevent 0 value

This commit fixes an issue where an empty string in the price conversion function resulted in a value of 0. Now, empty strings are properly handled and do not convert to a numerical value.

imanish003 commented:
@kmanijak JFYI Fixed the issue in [d47dadd](https://github.com/woocommerce/woocommerce/pull/47354/commits/d47dadd1201fa39c23042ec34d33608a2ccdcdaf) because of which an empty string was getting converted to 0.
